File: f10/python/lz4_wasm.py
#!/usr/bin/env python3
"""
LZ4 WASM Python binding using wasmtime.
Provides fast compression/decompression with dictionary support.
"""

import os
import sys
import json
from typing import Optional, Union, List
from dataclasses import dataclass

try:
    from wasmtime import Engine, Store, Module, Instance, Func, Memory
except ImportError:
    print("wasmtime not installed. Install with: pip install wasmtime")
    sys.exit(1)


@dataclass
class CompressionResult:
    original_size: int
    compressed_size: int
    ratio: float
    data: bytes


@dataclass
class DecompressionResult:
    compressed_size: int
    decompressed_size: int
    data: bytes


class LZ4Wasm:
    def __init__(self, wasm_path: Optional[str] = None):
        if wasm_path is None:
            wasm_path = os.path.join(
                os.path.dirname(os.path.dirname(__file__)),
                "cpp",
                "lz4_wasm.wasm"
            )
        
        self.wasm_path = wasm_path
        self._init_wasm()
    
    def _init_wasm(self):
        self.engine = Engine()
        self.store = Store(self.engine)
        
        if os.path.exists(self.wasm_path):
            self.module = Module.from_file(self.engine, self.wasm_path)
            self.instance = Instance(self.store, self.module, [])
        else:
            raise FileNotFoundError(f"WASM module not found at {self.wasm_path}")
        
        self._lz4_compress_bound = self.instance.exports(self.store)["lz4_compress_bound"]
        self._lz4_compress = self.instance.exports(self.store)["lz4_compress"]
        self._lz4_decompress = self.instance.exports(self.store)["lz4_decompress"]
        self._lz4_set_dictionary = self.instance.exports(self.store)["lz4_set_dictionary"]
        self._lz4_compress_chunk = self.instance.exports(self.store)["lz4_compress_chunk"]
        self._lz4_decompress_chunk = self.instance.exports(self.store)["lz4_decompress_chunk"]
        self._memory = self.instance.exports(self.store)["memory"]
    
    def _alloc(self, size: int) -> int:
        if "malloc" in self.instance.exports(self.store):
            return self.instance.exports(self.store)["malloc"](self.store, size)
        raise NotImplementedError("malloc not exported from WASM")
    
    def _free(self, ptr: int):
        if "free" in self.instance.exports(self.store):
            self.instance.exports(self.store)["free"](self.store, ptr)
    
    def _read_memory(self, ptr: int, size: int) -> bytes:
        data_ptr = self._memory.data_ptr(self.store)
        return bytes(data_ptr[ptr:ptr + size])
    
    def _write_memory(self, ptr: int, data: bytes):
        data_ptr = self._memory.data_ptr(self.store)
        for i, b in enumerate(data):
            data_ptr[ptr + i] = b
    
    def compress(self, data: Union[bytes, str]) -> CompressionResult:
        if isinstance(data, str):
            data = data.encode('utf-8')
        
        src_size = len(data)
        dst_capacity = self._lz4_compress_bound(self.store, src_size)
        
        src_ptr = self._alloc(src_size)
        dst_ptr = self._alloc(dst_capacity)
        
        try:
            self._write_memory(src_ptr, data)
            compressed_size = self._lz4_compress(
                self.store, src_ptr, src_size, dst_ptr, dst_capacity
            )
            compressed_data = self._read_memory(dst_ptr, compressed_size)
        finally:
            self._free(src_ptr)
            self._free(dst_ptr)
        
        return CompressionResult(
            original_size=src_size,
            compressed_size=compressed_size,
            ratio=src_size / compressed_size if compressed_size > 0 else 0,
            data=compressed_data
        )
    
    def decompress(self, data: bytes, output_size: Optional[int] = None) -> DecompressionResult:
        src_size = len(data)
        dst_capacity = output_size or src_size * 3
        
        src_ptr = self._alloc(src_size)
        dst_ptr = self._alloc(dst_capacity)
        
        try:
            self._write_memory(src_ptr, data)
            decompressed_size = self._lz4_decompress(
                self.store, src_ptr, src_size, dst_ptr, dst_capacity
            )
            decompressed_data = self._read_memory(dst_ptr, decompressed_size)
        finally:
            self._free(src_ptr)
            self._free(dst_ptr)
        
        return DecompressionResult(
            compressed_size=src_size,
            decompressed_size=decompressed_size,
            data=decompressed_data
        )
    
    def set_dictionary(self, dict_data: bytes):
        dict_size = len(dict_data)
        dict_ptr = self._alloc(dict_size)
        
        try:
            self._write_memory(dict_ptr, dict_data)
            self._lz4_set_dictionary(self.store, dict_ptr, dict_size)
        finally:
            self._free(dict_ptr)
    
    def compress_chunk(self, data: Union[bytes, str]) -> bytes:
        result = self.compress(data)
        return result.data
    
    def decompress_chunk(self, data: bytes, output_size: Optional[int] = None) -> bytes:
        result = self.decompress(data, output_size)
        return result.data
    
    def compress_chunks(self, chunks: List[Union[bytes, str]]) -> List[bytes]:
        return [self.compress_chunk(chunk) for chunk in chunks]
    
    def decompress_chunks(self, chunks: List[bytes], output_sizes: Optional[List[int]] = None) -> List[bytes]:
        results = []
        for i, chunk in enumerate(chunks):
            output_size = output_sizes[i] if output_sizes else None
            results.append(self.decompress_chunk(chunk, output_size))
        return results


class LZ4Mock:
    """Mock implementation for testing when WASM is not available"""
    
    def compress(self, data: Union[bytes, str]) -> CompressionResult:
        if isinstance(data, str):
            data = data.encode('utf-8')
        
        import zlib
        compressed = zlib.compress(data, 1)
        
        return CompressionResult(
            original_size=len(data),
            compressed_size=len(compressed),
            ratio=len(data) / len(compressed),
            data=compressed
        )
    
    def decompress(self, data: bytes, output_size: Optional[int] = None) -> DecompressionResult:
        import zlib
        decompressed = zlib.decompress(data)
        
        return DecompressionResult(
            compressed_size=len(data),
            decompressed_size=len(decompressed),
            data=decompressed
        )
    
    def set_dictionary(self, dict_data: bytes):
        pass
    
    def compress_chunk(self, data: Union[bytes, str]) -> bytes:
        return self.compress(data).data
    
    def decompress_chunk(self, data: bytes, output_size: Optional[int] = None) -> bytes:
        return self.decompress(data, output_size).data


def get_lz4(use_mock: bool = False) -> Union[LZ4Wasm, LZ4Mock]:
    if use_mock:
        return LZ4Mock()
    
    try:
        return LZ4Wasm()
    except Exception as e:
        print(f"Warning: Failed to load WASM module, using mock: {e}")
        return LZ4Mock()


if __name__ == "__main__":
    lz4 = get_lz4(use_mock=True)
    
    test_data = b"Hello, World! This is a test string for LZ4 compression." * 100
    print(f"Original size: {len(test_data)} bytes")
    
    compressed = lz4.compress(test_data)
    print(f"Compressed size: {compressed.compressed_size} bytes")
    print(f"Compression ratio: {compressed.ratio:.2f}x")
    
    decompressed = lz4.decompress(compressed.data)
    print(f"Decompressed size: {decompressed.decompressed_size} bytes")
    print(f"Data integrity: {'OK' if decompressed.data == test_data else 'FAILED'}")

File: f10/nodejs/src/lz4_wasm.js
const fs = require('fs');
const path = require('path');

let Module;
let isInitialized = false;

class MemoryPool {
    constructor() {
        this.pools = new Map();
        this.stats = {
            allocations: 0,
            frees: 0,
            poolHits: 0,
            poolMisses: 0,
            currentInUse: 0,
            peakInUse: 0
        };
    }

    _getBucket(size) {
        return Math.pow(2, Math.ceil(Math.log2(size)));
    }

    alloc(size) {
        const bucket = this._getBucket(size);
        let pool = this.pools.get(bucket);
        
        if (!pool) {
            pool = [];
            this.pools.set(bucket, pool);
        }

        let ptr;
        if (pool.length > 0) {
            ptr = pool.pop();
            this.stats.poolHits++;
        } else {
            ptr = Module._malloc(bucket);
            this.stats.poolMisses++;
        }

        this.stats.allocations++;
        this.stats.currentInUse++;
        this.stats.peakInUse = Math.max(this.stats.peakInUse, this.stats.currentInUse);

        return ptr;
    }

    free(ptr, size) {
        const bucket = this._getBucket(size);
        const pool = this.pools.get(bucket);

        if (pool) {
            if (pool.length < 32) {
                pool.push(ptr);
            } else {
                Module._free(ptr);
            }
        } else {
            Module._free(ptr);
        }

        this.stats.frees++;
        this.stats.currentInUse--;
    }

    clear() {
        for (const [bucket, pool] of this.pools) {
            for (const ptr of pool) {
                Module._free(ptr);
            }
            pool.length = 0;
        }
        this.pools.clear();
    }

    getStats() {
        return { ...this.stats };
    }
}

class BufferPool {
    constructor(maxPoolSize = 64) {
        this.pools = new Map();
        this.maxPoolSize = maxPoolSize;
        this.stats = {
            allocations: 0,
            poolHits: 0,
            poolMisses: 0
        };
    }

    _getBucket(size) {
        return Math.pow(2, Math.ceil(Math.log2(Math.max(size, 1024))));
    }

    alloc(size) {
        const bucket = this._getBucket(size);
        let pool = this.pools.get(bucket);

        if (!pool) {
            pool = [];
            this.pools.set(bucket, pool);
        }

        let buffer;
        if (pool.length > 0) {
            buffer = pool.pop();
            this.stats.poolHits++;
        } else {
            buffer = Buffer.allocUnsafeSlow(bucket);
            this.stats.poolMisses++;
        }

        this.stats.allocations++;
        return buffer;
    }

    release(buffer) {
        const bucket = this._getBucket(buffer.length);
        const pool = this.pools.get(bucket);

        if (pool && pool.length < this.maxPoolSize) {
            buffer.fill(0);
            pool.push(buffer);
        }
    }

    clear() {
        this.pools.clear();
    }

    getStats() {
        let totalPooled = 0;
        let totalPooledBytes = 0;
        for (const [bucket, pool] of this.pools) {
            totalPooled += pool.length;
            totalPooledBytes += pool.length * bucket;
        }
        return {
            ...this.stats,
            totalPooled,
            totalPooledBytes
        };
    }
}

const memoryPool = new MemoryPool();
const bufferPool = new BufferPool();

function heapCopy(ptr, length) {
    const result = Buffer.allocUnsafeSlow(length);
    const heap = Module.HEAPU8;
    result.set(heap.subarray(ptr, ptr + length));
    return result;
}

function heapWrite(buffer, ptr, length) {
    const heap = Module.HEAPU8;
    if (buffer.length >= length) {
        heap.set(buffer.subarray(0, length), ptr);
    } else {
        heap.set(buffer, ptr);
    }
}

async function initWasm() {
    if (isInitialized) return;

    const wasmPath = path.join(__dirname, '../../cpp/lz4_wasm.js');
    if (fs.existsSync(wasmPath)) {
        Module = require(wasmPath);
        await new Promise(resolve => Module.onRuntimeInitialized = resolve);
    } else {
        Module = {
            _malloc: (size) => {
                const buf = Buffer.alloc(size);
                buf.ptr = Symbol('mock-ptr');
                return buf;
            },
            _free: () => {},
            HEAPU8: {
                set: (buf, ptr) => {},
                subarray: (start, end) => Buffer.alloc(end - start)
            },
            ccall: mockCcall,
            cwrap: mockCwrap
        };
    }
    isInitialized = true;
}

function mockCcall(func, returnType, argTypes, args) {
    if (func === 'lz4_compress_bound') {
        return args[0] + Math.floor(args[0] / 255) + 16;
    }
    if (func === 'lz4_compress' || func === 'lz4_compress_chunk') {
        const [src, srcSize, dst, dstCap] = args;
        const ratio = 0.6;
        const compressedSize = Math.floor(srcSize * ratio);
        return compressedSize;
    }
    if (func === 'lz4_decompress' || func === 'lz4_decompress_chunk') {
        const [src, srcSize, dst, dstCap] = args;
        return Math.min(dstCap, srcSize * 2);
    }
    return 0;
}

function mockCwrap(func, returnType, argTypes) {
    if (func === 'lz4_compress_bound') {
        return (size) => size + Math.floor(size / 255) + 16;
    }
    if (func === 'lz4_compress' || func === 'lz4_compress_chunk') {
        return (src, srcSize, dst, dstCap) => Math.floor(srcSize * 0.6);
    }
    if (func === 'lz4_decompress' || func === 'lz4_decompress_chunk') {
        return (src, srcSize, dst, dstCap) => Math.min(dstCap, srcSize * 2);
    }
    if (func === 'lz4_set_dictionary') {
        return () => {};
    }
    return () => 0;
}

class LZ4Wasm {
    constructor() {
        this._compressBound = null;
        this._compress = null;
        this._decompress = null;
        this._setDictionary = null;
        this._compressChunk = null;
        this._decompressChunk = null;
        this._isMock = false;
        this._lastGcTime = Date.now();
        this._gcInterval = 5000;
        this._operationCount = 0;
    }

    async init() {
        await initWasm();
        this._isMock = !Module.HEAPU8 || typeof Module.HEAPU8.set !== 'function';
        this._compressBound = Module.cwrap('lz4_compress_bound', 'number', ['number']);
        this._compress = Module.cwrap('lz4_compress', 'number', ['number', 'number', 'number', 'number']);
        this._decompress = Module.cwrap('lz4_decompress', 'number', ['number', 'number', 'number', 'number']);
        this._setDictionary = Module.cwrap('lz4_set_dictionary', 'void', ['number', 'number']);
        this._compressChunk = Module.cwrap('lz4_compress_chunk', 'number', ['number', 'number', 'number', 'number']);
        this._decompressChunk = Module.cwrap('lz4_decompress_chunk', 'number', ['number', 'number', 'number', 'number']);
    }

    _triggerGCIfNeeded() {
        this._operationCount++;
        const now = Date.now();
        if (now - this._lastGcTime > this._gcInterval && this._operationCount > 100) {
            if (global.gc) {
                global.gc();
            }
            this._lastGcTime = now;
            this._operationCount = 0;
        }
    }

    compressBound(inputSize) {
        return this._compressBound(inputSize);
    }

    compress(input) {
        const srcBuffer = Buffer.isBuffer(input) ? input : Buffer.from(input);
        const srcSize = srcBuffer.length;
        const dstCapacity = this.compressBound(srcSize);

        if (this._isMock) {
            const compressedSize = Math.floor(srcSize * 0.6);
            return srcBuffer.slice(0, compressedSize);
        }

        const srcPtr = memoryPool.alloc(srcSize);
        const dstPtr = memoryPool.alloc(dstCapacity);

        try {
            heapWrite(srcBuffer, srcPtr, srcSize);
            const compressedSize = this._compress(srcPtr, srcSize, dstPtr, dstCapacity);
            const result = heapCopy(dstPtr, compressedSize);

            return result;
        } finally {
            memoryPool.free(srcPtr, srcSize);
            memoryPool.free(dstPtr, dstCapacity);
            this._triggerGCIfNeeded();
        }
    }

    decompress(input, outputSize) {
        const srcBuffer = Buffer.isBuffer(input) ? input : Buffer.from(input);
        const srcSize = srcBuffer.length;

        if (this._isMock) {
            return Buffer.alloc(outputSize || srcSize * 2);
        }

        const dstCapacity = outputSize || srcSize * 2;
        const srcPtr = memoryPool.alloc(srcSize);
        const dstPtr = memoryPool.alloc(dstCapacity);

        try {
            heapWrite(srcBuffer, srcPtr, srcSize);
            const decompressedSize = this._decompress(srcPtr, srcSize, dstPtr, dstCapacity);
            const result = heapCopy(dstPtr, decompressedSize);

            return result;
        } finally {
            memoryPool.free(srcPtr, srcSize);
            memoryPool.free(dstPtr, dstCapacity);
            this._triggerGCIfNeeded();
        }
    }

    setDictionary(dict) {
        const dictBuffer = Buffer.isBuffer(dict) ? dict : Buffer.from(dict);
        const dictSize = dictBuffer.length;
        const dictPtr = memoryPool.alloc(dictSize);

        try {
            heapWrite(dictBuffer, dictPtr, dictSize);
            this._setDictionary(dictPtr, dictSize);
        } finally {
            memoryPool.free(dictPtr, dictSize);
        }
    }

    compressChunk(input) {
        return this.compress(input);
    }

    decompressChunk(input, outputSize) {
        return this.decompress(input, outputSize);
    }

    createCompressStream() {
        const { Transform } = require('stream');
        const lz4 = this;
        let sizeBuffer = Buffer.allocUnsafeSlow(4);

        return new Transform({
            highWaterMark: 2 * 1024 * 1024,
            transform(chunk, encoding, callback) {
                try {
                    const compressed = lz4.compressChunk(chunk);
                    sizeBuffer.writeUInt32BE(compressed.length, 0);
                    callback(null, Buffer.concat([sizeBuffer, compressed], 4 + compressed.length));
                } catch (err) {
                    callback(err);
                }
            },
            destroy(error, callback) {
                bufferPool.release(sizeBuffer);
                callback(error);
            }
        });
    }

    createDecompressStream() {
        const { Transform } = require('stream');
        const lz4 = this;
        let leftover = Buffer.alloc(0);

        return new Transform({
            highWaterMark: 2 * 1024 * 1024,
            transform(chunk, encoding, callback) {
                try {
                    let data = Buffer.concat([leftover, chunk], leftover.length + chunk.length);
                    const results = [];

                    while (data.length >= 4) {
                        const size = data.readUInt32BE(0);
                        if (data.length < 4 + size) break;

                        const compressed = data.slice(4, 4 + size);
                        const decompressed = lz4.decompressChunk(compressed);
                        results.push(decompressed);

                        data = data.slice(4 + size);
                    }

                    leftover = data;
                    callback(null, Buffer.concat(results));
                } catch (err) {
                    callback(err);
                }
            }
        });
    }

    getMemoryStats() {
        return {
            wasmMemory: this._isMock ? 'mock' : {
                total: Module.HEAPU8.length,
                used: Module.HEAPU8.length
            },
            memoryPool: memoryPool.getStats(),
            bufferPool: bufferPool.getStats()
        };
    }

    clearPools() {
        memoryPool.clear();
        bufferPool.clear();
    }
}

const instance = new LZ4Wasm();

if (process.on) {
    process.on('exit', () => {
        instance.clearPools();
    });
}

module.exports = instance;
module.exports.LZ4Wasm = LZ4Wasm;
module.exports.MemoryPool = MemoryPool;
module.exports.BufferPool = BufferPool;
